src/v/config/property_constraints.h

```diff
@@ -0,0 +1,268 @@
+        T topic_val = value_extractor(topic_cfg);
+        if (min) {
+            topic_val = std::max(*min, topic_val);
+        }
+        if (max) {
+            topic_val = std::min(*max, topic_val);
+        }
+        clamper(topic_cfg, topic_val);
+    }
+
+    void encode_yaml(YAML::Node& n) const override {
+        if (min) {
+            n["min"] = *min;
+        }
+
+        if (max) {
+            n["max"] = *max;
+        }
+    }
+
+    void rjson_serialize(json::Writer<json::StringBuffer>& w) const override {
+        w.Key("min");
+        if (min) {
+            rjson_serialize_integral<T>(w, min);
+        } else {
+            w.Null();
+        }
+
+        w.Key("max");
+        if (max) {
+            rjson_serialize_integral<T>(w, max);
+        } else {
+            w.Null();
+        }
+    }
+
+    void print(std::ostream& os) const override {
+        os << fmt::format("min: {}", min);
+        os << fmt::format("max: {}", max);
+    }
+};
+
+template<typename T>
+struct constraint_validator_enabled : public constraint_validator_t {
+    using extractor_t =
+      typename ss::noncopyable_function<T(const cluster::topic_configuration&)>;
+    using clamper_t = typename ss::noncopyable_function<void(
+      cluster::topic_configuration&, T&)>;
+
+    constraint_enabled_t enabled;
+    binding<T> original_property;
+    extractor_t value_extractor;
+    clamper_t clamper;
+
+    constraint_validator_enabled(
+      constraint_enabled_t enabled,
+      binding<T> original,
+      extractor_t value_extractor,
+      clamper_t clamper)
+      : enabled{std::move(enabled)}
+      , original_property{std::move(original)}
+      , value_extractor{std::move(value_extractor)}
+      , clamper{std::move(clamper)} {}
+
+    bool
+    validate(const cluster::topic_configuration& topic_cfg) const override {
+        T topic_val = value_extractor(topic_cfg);
+        return topic_val == original_property();
+    }
+
+    void clamp(cluster::topic_configuration& topic_cfg) const override {
+        T topic_val = value_extractor(topic_cfg);
+        topic_val = original_property();
+        clamper(topic_cfg, topic_val);
+    }
+
+    void encode_yaml(YAML::Node& n) const override {
+        n["enabled"] = enabled == constraint_enabled_t::yes ? true : false;
+    }
+
+    void rjson_serialize(json::Writer<json::StringBuffer>& w) const override {
+        w.Key("enabled");
+        w.Bool(enabled == constraint_enabled_t::yes);
+    }
+
+    void print(std::ostream& os) const override {
+        os << "enabled: " << enabled;
+    }
+};
+
+enum class constraint_type {
+    none = 0,
+    _restrict = 1,
+    clamp = 2,
+};
+
+std::string_view to_string_view(constraint_type type);
+
+template<>
+std::optional<constraint_type>
+from_string_view<constraint_type>(std::string_view sv);
+
+struct constraint_t {
+    ss::sstring name;
+    constraint_type type;
+    std::shared_ptr<constraint_validator_t> validator;
```

Comment:
> Also why not `ss::shared_ptr`? (or the light weight one)

Just wanted to see if you would catch this ;) 

> question: what's your reasoning for the validator to be a `shared_ptr`? Why doesn't `unique_ptr` work?

I'll try to keep this short and we can save longer talk for voice. Our property classes currently make copies of the type in two places, at the constructor for the modifiable value and assignment operator which is a problem with unique_ptr. There were two ways to address this:
1. Add copy methods to the constraint objects that make new unique_ptrs - this is similar to what a shared_ptr can do, so why use a unique_ptr?
2. Create a new property class - This would move separate instances of `vector<constraint_t>` into the modifiable value and default value at the constructor. This would also require an overload for the assignment operator. A consequence of this design is when the property is reset to default state, the default value must be moved and then re-created. Another consequence is that several other methods from the base_property interface needed overloading to support moving. 

Point 2 seemed like a lot of code change for something this small, so I opted for point 1 for now.

Finally, at it's heart, the pointer design is to support creating constraints at runtime but constraints could be formed at compile time. Thus I created the follow up ticket https://github.com/redpanda-data/redpanda/issues/14052 to re-structure the architecture and remove the need for these pointers entirely.